crates/goose/src/agents/truncate.rs

```
@@ -246,152 +101,170 @@ impl Agent for TruncateAgent {
+                            );
+                        }
+
+                        yield message_tool_response.clone();
+
+                        messages.push(response);
+                        messages.push(message_tool_response);
+                    },
+                    Err(ProviderError::ContextLengthExceeded(_)) => {
+                        if truncation_attempt >= MAX_TRUNCATION_ATTEMPTS {
+                            // Create a user-facing error message & terminate the stream
+                            let error_message = Message::user().with_text("Error: Context length exceeds limits even after multiple attempts to truncate.");
+                            yield error_message;
+                            break;
+                        }
+
+                        truncation_attempt += 1;
+                        debug!("Context length exceeded. Initiating truncation attempt {}/{}.", truncation_attempt, MAX_TRUNCATION_ATTEMPTS);
+
+                        // Decay the estimate factor as we make more truncation attempts
+                        // Estimate factor decays like this over time: 0.9, 0.81, 0.729, ...
+                        let estimate_factor: f32 = ESTIMATE_FACTOR_DECAY.powi(truncation_attempt as i32);
+
+                        // release the lock before truncation to prevent deadlock
+                        drop(capabilities);
+
+                        self.truncate_messages(&mut messages, estimate_factor).await?;
+
+                        // Re-acquire the lock
+                        capabilities = self.capabilities.lock().await;
+
+                        // Retry the loop after truncation
+                        continue;
+                    },
+                    Err(e) => {
+                        // Create a user-facing error message & terminate the stream
+                        error!("Error: {}", e);
+                        let error_message = Message::user().with_text(format!("Error: {}", e));
```

i think the previous message would have been a user message (e.g. before any tool calls, this is just after the input message. at the start of a loop after a tool call, it would be after a tool_use assistant followed by a tool_result user). so we need to surface an assistant error here instead if we treat it as a message?